Maintain dirtiness and use retained engine layers

liyuqian opened the pull request:
flutter/flutter#21756

liyuqian committed:
Maintain dirtiness and use retained engine layers

liyuqian commented:
@Hixie as I'm waiting for Chinmay to review https://github.com/flutter/engine/pull/6468 and https://github.com/flutter/engine/pull/6442, I think maybe you can start reviewing the corresponding framework change here.

All the tests would fail until those engine commits get merged. So please ignore them, and the dependency overrides or `//TODO TEST` debugging pieces. Maybe just focusing on the dirtiness and the proof of the non-existence of the loop?

liyuqian committed:
test

liyuqian committed:
Keep LeaderLayer and FollowerLayer dirty

liyuqian committed:
Update subtree dirtiness for widget inspector and screenshot

liyuqian committed:
Remove test code

liyuqian commented:
@Hixie this patch is ready to review as the engine commits have all rolled into framework and now all tests have been green for this patch.

Hixie commented on packages/flutter/lib/src/material/page_transitions_theme.dart:
```diff
@@ -37,9 +37,15 @@ class _FadeUpwardsPageTransition extends StatelessWidget {
+        // The above RepaintBoundary separates slide animation from fade
+        // animation so we can apply retained rendering optimization to the fade
+        // animation. Otherwise, the offset change from slide will mix with the
+        // opacity change and our optimization won't be applied.
```
this is very unintuitive. How are our users supposed to know this happens?

Hixie replied:
```suggestion
        // This RepaintBoundary separates the slide animation from the fade
        // animation so the retained rendering optimization maintains the
        // identity of the opacity layer from the fade animation. Without this
        // boundary, the offset change from the slide animation would be applied
        // to the opacity layer and the layer would not be retained.
```

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -41,6 +41,33 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   @override
   ContainerLayer get parent => super.parent;
 
+  // Whether this layer has any changes since it's last added to the engine
```
```suggestion
  // Whether this layer has any changes since it was last sent to the engine
```

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -41,6 +41,33 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+
+  /// Mark that this layer has changed and we need to upload changes to the
+  /// engine using [addToScene].
+  void markDirty() { _isDirty = true; }
```
we call these things markNeedsXXX in Flutter. e.g. markNeedsUpdate

Hixie replied:
but i guess markDirty is fine

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -41,6 +41,33 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+    if (this is ContainerLayer) {
+      final ContainerLayer container = this;
+      for (Layer child = container.firstChild; child != null; child = child.nextSibling) {
+        child.updateSubtreeDirtiness();
```
ah, i see that we do, because you sometimes mark entire parts of the subtree as clean but don't clear this flag at the same time

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -463,9 +524,23 @@ class ContainerLayer extends Layer {
+          'WARNING: please use OffsetLayer to handle offset and do not '
+          'pass offsets down to children as it disables our retained '
+          'rendering optimization.'
+        );
```
please either fail (throw a FlutterError; see https://master-docs-flutter-io.firebaseapp.com/flutter/foundation/FlutterError/message.html for details), or don't output anything. We don't do warnings, because people ignore them and you end up with lots of stuff in the logs making development harder.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -564,15 +639,16 @@ class OffsetLayer extends ContainerLayer {
+    final ui.EngineLayer engineLayer = builder.pushOffset(layerOffset.dx + offset.dx, layerOffset.dy + offset.dy);
     addChildrenToScene(builder);
     builder.pop();
+    return engineLayer; // this does not return an engine layer yet.
```
the comment seems to disagree with the code?

liyuqian replied:
Fixed.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1348,14 +1435,14 @@ class FollowerLayer extends ContainerLayer {
       _lastOffset = null;
       _inverseDirty = true;
-      return;
+      return null;
```
you sometimes have the "this does not have an engine layer" comment and sometimes not, which seems confusing

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -198,6 +198,9 @@ class PaintingContext extends ClipContext {
         return true;
       }());
     }
+    if (child._layer.offset != offset) {
```
```suggestion
    assert(child._layer != null);
    if (child._layer.offset != offset) {
```

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -41,6 +41,33 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+        _isSubtreeDirty = _isSubtreeDirty || child._isSubtreeDirty;
+      }
+    }
+  }
```
oh i see, it's not O(N^2) because you only run this once, before sending the layers up to the engine.

I recommend making this private, and having a single method that does all the logic of updating these bits, creating the SceneBuilder, calling addToScene, and eventually returning a Scene. That way the two places where you call this method would just call that method and there would be less danger of someone thinking they had to call this method sometimes. (That would also allow you to make addToScene protected.)

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -104,7 +131,35 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// corresponding engine layer, null is returned.
+  ui.EngineLayer addToScene(ui.SceneBuilder builder, [Offset layerOffset = Offset.zero]);
+
+  void _addToSceneWithRetainedRendering(ui.SceneBuilder builder) {
```
how do i call this if my layer isn't defined in this file?

Hixie commented on packages/flutter/lib/src/material/page_transitions_theme.dart:
```diff
@@ -37,9 +37,15 @@ class _FadeUpwardsPageTransition extends StatelessWidget {
+        ),
       ),
     );
   }
```
I don't understand how the OpacityLayer gets retained. Every time the RenderOpacity object paints, i.e. every frame of the fade animation, the OpacityLayer is recreated, which means the engine layer is lost. no?

Hixie committed:
Update packages/flutter/lib/src/material/page_transitions_theme.dart

Hixie committed:
Update packages/flutter/lib/src/rendering/layer.dart

Hixie committed:
Update packages/flutter/lib/src/rendering/layer.dart

Hixie committed:
Update packages/flutter/lib/src/rendering/object.dart

liyuqian commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -104,7 +131,35 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+      return;
+    }
+    _engineLayer = addToScene(builder);
+    if (this is FollowerLayer || this is LeaderLayer) {
```
Done.

liyuqian committed:
Fixes

liyuqian reviewed (commented):
Thanks for the comments! Fixed accordingly and I'll think more about how to encapsulate addToScene and updateSubtreeDirtiness.

liyuqian committed:
Make addToScene & updateSubtreeDirtiness protected

We make a new API buildScene to encapsulate those calls.

liyuqian committed:
Capture offset in OpacityLayer

liyuqian committed:
Add unit tests

liyuqian committed:
Remove test code

liyuqian commented:
@Hixie : I've added unit tests, removed the additional RepaintBoundary, and let OpacityLayer catch the offset change. Now it's ready for another round of review.

liyuqian committed:
Suppress invalid_use_of_protected_member in tests

liyuqian commented:
Github code review is broken for this PR...
